[Clang][Sema] Diagnose use of template keyword after declarative nested-name-specifiers

sdkrystian opened the pull request:
According to [[temp.names] p5](http://eel.is/c++draft/temp.names#5):
> The keyword template shall not appear immediately after a declarative [nested-name-specifier](http://eel.is/c++draft/expr.prim.id.qual#nt:nested-name-specifier).

[[expr.prim.id.qual] p2](http://eel.is/c++draft/expr.prim.id.qual#2) defines a declarative [nested-name-specifier](http://eel.is/c++draft/expr.prim.id.qual#nt:nested-name-specifier) as follows:
> A [nested-name-specifier](http://eel.is/c++draft/expr.prim.id.qual#nt:nested-name-specifier) is [declarative](http://eel.is/c++draft/expr.prim.id.qual#def:declarative) if it is part of
> - a [class-head-name](http://eel.is/c++draft/class.pre#nt:class-head-name),
> - an [enum-head-name](http://eel.is/c++draft/dcl.enum#nt:enum-head-name),
> - a [qualified-id](http://eel.is/c++draft/expr.prim.id.qual#nt:qualified-id) that is the [id-expression](http://eel.is/c++draft/expr.prim.id.general#nt:id-expression) of a [declarator-id](http://eel.is/c++draft/dcl.decl.general#nt:declarator-id), or
> - a declarative [nested-name-specifier](http://eel.is/c++draft/expr.prim.id.qual#nt:nested-name-specifier)[.](http://eel.is/c++draft/expr.prim.id.qual#2.sentence-1)

Note: I believe this definition is defective as it doesn't include _nested-name-specifiers_ appearing in _elaborated-type-specifiers_ that declare partial/explicit specializations and explicit instantiations. See my post to the core reflector [here](https://lists.isocpp.org/core/2024/01/15325.php). Minus a few bugs that are addressed by this PR, this is how we implement it.

This means that declarations like:
```cpp
template<typename>
struct A
{
    template<typename> 
    struct B
   {
        void f();
   };
};

template<typename T>
template<typename U>
void A<T>::template B<U>::f() { } // error: 'template' cannot be used after a declarative nested name specifier
```
are ill-formed. This PR add diagnostics for such declarations. The name of the diagnostic group is `template-in-declaration-name`.

Regarding the aforementioned "few bugs that are addressed by this PR" in order to correctly implement this:
- `CheckClassTemplate` did not call `diagnoseQualifiedDeclaration` when the semantic context was dependent. This allowed for constructs like:
```cpp
struct A
{
    template<typename T>
    struct B
    {
        template<typename U>
        struct C;
    };
};

template<typename T>
template<typename U>
struct decltype(A())::B<T>::C { };
```
- `ActOnClassTemplateSpecialization` did not call `diagnoseQualifiedDeclaration` at all, allowing for qualified partial/explicit specializations at class scope and other related nonsense
- `TreeTransform::TransformNestedNameSpecifierLoc` would rebuild a `NestedNameSpecifier::TypeSpecWithTemplate` as a `NestedNameSpecifier::TypeSpec`
- `TemplateSpecializationTypeLoc::initializeLocal` would set the `template` keyword `SourceLocation` to the provided `Loc` parameter, which would result in a `TemplateSpecializationTypeLoc` obtained via `ASTContext::getTrivialTypeSourceInfo` being displayed as always having a `template` prefix (since the presence of the keyword is not stored anywhere else).

sdkrystian committed:
[Clang][Sema] Diagnose use of template keyword after declarative nested-name-specifiers

sdkrystian commented:
(still need to add a release note)

Endilll commented on clang/test/CXX/drs/dr23xx.cpp:
```diff
@@ -182,8 +182,8 @@ struct Bad2 { int a, b; };
 } // namespace dr2386
 namespace std {
 template <typename T> struct tuple_size;
-template <> struct std::tuple_size<dr2386::Bad1> {};
```
Is this a really necessary change?

sdkrystian replied:
@Endilll if left unchanged this will result in a warning about redundant qualification, which I don't think is the point of the test... alternatively I can add `// expected-warning {{extra qualification on member}}`

Endilll replied:
Is this warning emitted if we spell it `::std::tuple)size`?

sdkrystian replied:
It is, yes

cor3ntin commented:
Thank for this patch 

I think this might be better done as a pedantic warning.
IE, I am concerned that this could break existing code, similarly to https://github.com/llvm/llvm-project/issues/78031#issuecomment-1891475606

I wonder if we should add tests for https://cplusplus.github.io/CWG/issues/1707.html in this patch.

sdkrystian commented:
@cor3ntin Pedantic warning would probably be best as other implementations are not great at diagnosing this

sdkrystian commented:
@cor3ntin I changed the diagnostic to a pedantic warning and added a release note. 

> I wonder if we should add tests for https://cplusplus.github.io/CWG/issues/1707.html in this patch.

We don't check for the template keyword when parsing an _elaborated-type-specifier_ without a _nested-name-specifier_, so such a declaration is parsed as an unnamed class. I feel that it's beyond the scope of this PR anyways, as this is only concerned with qualified declarations.

cor3ntin commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -4394,8 +4394,14 @@ NestedNameSpecifierLoc TreeTransform<Derived>::TransformNestedNameSpecifierLoc(
         }
-        SS.Extend(SemaRef.Context, /*FIXME:*/ SourceLocation(), TL,
-                  Q.getLocalEndLoc());
+        SourceLocation TemplateKWLoc;
+        if (const auto TSTL = TL.getAs<TemplateSpecializationTypeLoc>())
+          TemplateKWLoc = TSTL.getTemplateKeywordLoc();
+        else if (const auto DTSTL =
+                     TL.getAs<DependentTemplateSpecializationTypeLoc>())
+          TemplateKWLoc = DTSTL.getTemplateKeywordLoc();
```
Yes, we should make a function :)

cor3ntin replied:
In TypeLoc sounds good to me

sdkrystian commented on clang/test/CXX/temp/temp.decls/temp.class/temp.mem.func/p1.cpp:
```diff
@@ -3,21 +3,21 @@ template<typename T, typename U> // expected-note{{previous template}}
-  
+
   X0(int);
   ~X0();
-  
+
   void f0(const T&, const U&);
-  
+
   T& operator[](int i) const;
-  
+
   void f1(size_type) const;
   void f2(size_type) const;
   void f3(size_type) const;
   void f4() ;
-  
+
   operator T*() const;
-  
+
   T value;
 };
 
```
Oh, whoops. Sure.

sdkrystian commented:
@cor3ntin Addressed your comments. I also added the error to a diagnostics group (`template-in-declaration-name`)

cor3ntin reviewed (approved):
LGTM, thanks!

sdkrystian commented:
@erichkeane cor3ntin approved... please take a look when you have time

rupprecht commented:
It looks like this caused some new `-Wextra-qualification` warnings (playground: https://godbolt.org/z/3MbMjGYET)

```
namespace foo {
template <typename T>
struct Z {};

template <>
struct Z<bool> {}; // OK

template <>
struct foo::Z<int> {}; // New warning: extra qualification on member 'Z' [-Wextra-qualification]

template <>
struct ::foo::Z<double> {}; // New warning: extra qualification on member 'Z' [-Wextra-qualification]
}  // namespace foo
```

This doesn't use `template` in the way described in the commit or the release notes, so it doesn't seem like an intentional change. But GCC already errors on this (and downgrades to a warning with `-fpermissive`), so Clang isn't necessarily wrong to start warning about this now. Still, it's possible GCC and Clang are both wrong now, so I just want to make sure that's actually a desired side effect of this commit.

This behavior change is covered by the test diffs here in clang/test/CXX/drs/dr23xx.cpp and clang/test/SemaTemplate/class-template-spec.cpp.

sdkrystian commented:
@rupprecht This is intended behavior. We previously did not check qualified declarations of explicit/partial specializations of class templates for redundant qualification & invalid nested-name-specifiers (e.g. ones that begin with `decltype`). We do now, hence the new warning.

rupprecht commented:
Ok, thanks for the quick confirmation!
